### PR TITLE
[batch] spending limit kill switch for a billing project

### PR DESCRIPTION
**Pull request by jigold**
Stacked on #9354

I'm happy with the tests here.

### CODE REVIEW

File: batch/batch/front_end/front_end.py [hunk 1]
```diff
@@ -1503,6 +1526,27 @@ async def delete_batch_loop_body(app):
...
+SELECT id, user
+FROM batches
+WHERE billing_project = %s AND state = 'running';
+''',
+                                                          (record['billing_project'],))
+                async for batch in running_batches:
+                    await _cancel_batch(app, batch['id'], batch['user'])
+
+        # what should the time be here? There's on average 3 batch front-end instances.
+        await asyncio.sleep(30)
```
Comment (danking) [inline batch/batch/front_end/front_end.py:1547]: This kind of feels like something the driver should do. Sometimes we scale up to ~10 front ends. We don't want to hammer the DB with 10 of these queries.

**Review by danking — DISMISSED** [review]

File: batch/test/test_batch.py [hunk 2]
```diff
@@ -704,3 +704,35 @@ def approx_equal(x, y, tolerance=1e-10):
...
+            j4 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j3])
+            j5 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j4])
+            j6 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j5])
+            j7 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j6])
+            j8 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j7])
+            j9 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j8])
+            j10 = batch.create_job('ubuntu:18.04', command=['sleep', '5'], parents=[j9])
+            batch = batch.submit()
+            batch = batch.wait()
+            assert batch['state'] == 'cancelled', batch
```
Comment (danking) [inline batch/test/test_batch.py:736]: I feel confident this test will fail, but I want something even stronger. If the cancel test runs every ten seconds, then I shouldn't overspend by, say, more than 30 seconds? Or at least one minute?
```python3
try:
    batch = client.create_batch()
    j1 = batch.create_job('ubuntu:18.04', command=['sleep', '5'])
    j2 = batch.create_job('ubuntu:18.04',
                          command=['/bin/bash', '-c', 'while true; do; sleep 3600; done'],
                          parents=[j1])
    batch = batch.submit()
    status = batch.wait()
    assert status['state'] == 'cancelled', batch
    jobs = {b['job_id']: b for b in batch.jobs()}
    assert jobs[2]['msec_mcpu'] < 30_000 * 125  # 30s on 125mCPU
```
Hmm. Oddly we don't have durations for cancelled jobs?

**Review by danking — CHANGES_REQUESTED** [review]
Looks great, just some organization and SLO things

File: batch/batch/front_end/front_end.py [hunk 3]
```diff
@@ -1320,58 +1408,12 @@ async def ui_get_billing(request, userdata):
...
-    if where_conditions:
-        where_condition = f'WHERE {" AND ".join(where_conditions)}'
-    else:
-        where_condition = ''
-
-    sql = f'''
-SELECT billing_projects.name as billing_project,
-    billing_projects.`status` as `status`,
-    users
-FROM (
```
Comment (jigold) [inline batch/batch/front_end/front_end.py:1345]: The rebase for this was horrendous. Can you double check the query is still correct in utils.py?

File: batch/batch/batch.py [hunk 4]
```diff
@@ -78,6 +80,28 @@ def _time_msecs_str(t):
     return d
 
 
+async def cancel_batch_in_db(db, batch_id, user):
+    @transaction(db)
+    async def cancel(tx):
+        record = await tx.execute_and_fetchone(
+            '''
+    SELECT `state` FROM batches
+    WHERE user = %s AND id = %s AND NOT deleted;
```
Comment (danking) [inline batch/batch/batch.py:89]: this needs a FOR UPDATE

**Review by danking — CHANGES_REQUESTED** [review]

File: batch/batch/utils.py [hunk 5]
```diff
@@ -158,3 +159,62 @@ def __repr__(self):
...
+) AS t
+RIGHT JOIN billing_projects
+  ON t.billing_project = billing_projects.name
+LEFT JOIN aggregated_billing_project_resources
+  ON aggregated_billing_project_resources.billing_project = billing_projects.name
+LEFT JOIN resources
+  ON resources.resource = aggregated_billing_project_resources.resource
+{where_condition}
+GROUP BY billing_projects.name, billing_projects.status, users, msec_mcpu, `limit`
+LOCK IN SHARE MODE;
```
Comment (jigold) [inline batch/batch/utils.py:200]: @danking I added a LOCK IN SHARE MODE here. Double check that please.

**Comment by jigold** [comment]
@danking  See comment. Otherwise, should be good to go.

**Review by danking — APPROVED** [review]
lgtm

**Pull request merged**